src/circom/mod.rs

```diff
@@ -0,0 +1,135 @@
+use std::path::PathBuf;
+
+use crate::{
+  errors::SpartanError,
+  traits::{self, snark::RelaxedR1CSSNARKTrait, Group},
+  ProverKey, VerifierKey, SNARK,
+};
+
+use bellpepper_core::{Circuit, ConstraintSystem, SynthesisError};
+use ff::PrimeField;
+
+use circom_scotia::{r1cs::R1CS, reader::load_r1cs, witness::WitnessCalculator};
+
+#[derive(Clone, Debug)]
+pub struct SpartanCircuit<F: PrimeField> {
+  r1cs: R1CS<F>,
+  witness: Option<Vec<F>>, // this is actually z = [1 || x || w]
+}
+
+#[allow(dead_code)]
+impl<F: PrimeField> SpartanCircuit<F> {
+  pub fn new(r1cs_path: PathBuf) -> Self {
+    SpartanCircuit {
+      r1cs: load_r1cs(r1cs_path),
+      witness: None,
+    }
+  }
+
+  pub fn compute_witness(&mut self, input: Vec<(String, Vec<F>)>, wtns_path: PathBuf) {
+    let mut witness_calculator = WitnessCalculator::new(wtns_path).unwrap();
+    let witness = witness_calculator
+      .calculate_witness(input.clone(), true)
+      .expect("msg");
+
+    self.witness = Some(witness);
+  }
+}
+
+impl<F: PrimeField> Circuit<F> for SpartanCircuit<F> {
+  fn synthesize<CS: ConstraintSystem<F>>(self, cs: &mut CS) -> Result<(), SynthesisError> {
+    let _ = circom_scotia::synthesize(cs, self.r1cs.clone(), self.witness).unwrap();
+
+    Ok(())
+  }
+}
+
+#[allow(dead_code)]
+pub fn setup<G: Group, S: RelaxedR1CSSNARKTrait<G>>(
+  circuit: SpartanCircuit<<G as Group>::Scalar>,
+) -> (ProverKey<G, S>, VerifierKey<G, S>) {
+  SNARK::<G, S, SpartanCircuit<<G as Group>::Scalar>>::setup(circuit).unwrap()
+}
+
+#[allow(dead_code)]
+pub fn generate_proof<G: Group, S: RelaxedR1CSSNARKTrait<G>>(
+  pk: ProverKey<G, S>,
+  circuit: &mut SpartanCircuit<<G as Group>::Scalar>,
+  input: Vec<(String, Vec<<G as Group>::Scalar>)>,
+  wtns_path: PathBuf,
+) -> Result<SNARK<G, S, SpartanCircuit<<G as traits::Group>::Scalar>>, SpartanError> {
+  circuit.compute_witness(input, wtns_path);
+  SNARK::prove(&pk, circuit.clone())
+}
+
+#[allow(dead_code)]
+pub fn create_snark<G: Group, S: RelaxedR1CSSNARKTrait<G>>(
```

Comment:
This method seems not that useful. In other words, we can just have two methods, one for setup and another for prove. The setup takes the r1cs file path and the prove takes the witness file path (in addition to pk).